src/gtest/test_proofs.cpp

```diff
@@ -0,0 +1,511 @@
+#include <gtest/gtest.h>
+#include "zcash/Proof.hpp"
+
+#include <iostream>
+
+#include "libsnark/common/default_types/r1cs_ppzksnark_pp.hpp"
+#include "libsnark/zk_proof_systems/ppzksnark/r1cs_ppzksnark/r1cs_ppzksnark.hpp"
+#include "zk_proof_systems/ppzksnark/r1cs_ppzksnark/r1cs_ppzksnark.hpp"
+#include "relations/constraint_satisfaction_problems/r1cs/examples/r1cs_examples.hpp"
+
+using namespace libzcash;
+
+typedef libsnark::default_r1cs_ppzksnark_pp curve_pp;
+typedef libsnark::default_r1cs_ppzksnark_pp::G1_type curve_G1;
+typedef libsnark::default_r1cs_ppzksnark_pp::G2_type curve_G2;
+typedef libsnark::default_r1cs_ppzksnark_pp::GT_type curve_GT;
+typedef libsnark::default_r1cs_ppzksnark_pp::Fp_type curve_Fr;
+typedef libsnark::default_r1cs_ppzksnark_pp::Fq_type curve_Fq;
+typedef libsnark::default_r1cs_ppzksnark_pp::Fqe_type curve_Fq2;
+
+#include "streams.h"
+#include "version.h"
+#include "utilstrencodings.h"
+
+TEST(proofs, sqrt_fq)
+{
+    // Poor man's PRNG
+    curve_Fq acc = curve_Fq("348957923485290374852379485") ^ 1000;
+
+    size_t quadratic_residues = 0;
+    size_t quadratic_nonresidues = 0;
+
+    for (size_t i = 1; i < 1000; i++) {
+        try {
+            acc += curve_Fq("45634563456") ^ i;
```

Comment:
This is only testing field elements in the range (348957923485290374852379485, 348957923485290374852379485 + 1000*45634564072], which is small compared to q=21888242871839275222246405745257275088696311157297823662689037894645226208583. I guess that's fine.